File: internal/eventlog/eventlog.go
// Package eventlog is the one place a tool call becomes a record
// (prd-event-log). The dispatcher emits exactly one Record per tools/call —
// success, refusal or error — and the auth layer contributes who made it.
//
// It is its own package rather than a corner of mcp because two layers write
// into it from opposite sides: auth knows the principal before the tool name
// exists, and mcp knows the tool and its outcome but never sees the
// credential. Neither may import the other, so the record's shape lives here
// and both depend on it.
package eventlog

import (
	"context"
	"log/slog"
	"time"
)

// Principal is who made the call, spelled the way AC1 asks: a JWT is its
// subject, an API key is its position in MCP_API_KEYS. The credential itself
// is never here — AC3 forbids it in any record, and the cheapest way to keep
// a value out of every record is to never let it into the type.
type Principal struct {
	// Method is how the caller authenticated: "jwt", "api_key", or "none"
	// when the deployment serves /mcp without authentication.
	Method string
	// ID is the subject (jwt) or the 1-based key index as text (api_key).
	ID string
}

// Anonymous is the principal of a deployment that authenticates nobody
// (MCP_AUTH_DISABLED). It is a value rather than a missing field so the
// principal column is never empty: an empty column cannot be told from a
// layer that forgot to fill it in.
var Anonymous = Principal{Method: "none"}

// Unauthenticated is the principal of a call auth refused (AC4). It is not
// Anonymous: that one passed an auth layer that asks nothing, this one failed
// an auth layer that asked. What it presented is not here either (AC3).
var Unauthenticated = Principal{Method: "unauthenticated"}

// String renders the principal for the record: "jwt:<sub>", "api_key:<n>",
// "anonymous" or "unauthenticated".
func (p Principal) String() string {
	switch p.Method {
	case "", "none":
		return "anonymous"
	case "unauthenticated":
		return "unauthenticated"
	default:
		return p.Method + ":" + p.ID
	}
}

type principalKey struct{}

// WithPrincipal stores the authenticated principal for the rest of the
// request.
func WithPrincipal(ctx context.Context, p Principal) context.Context {
	return context.WithValue(ctx, principalKey{}, p)
}

// PrincipalFrom reads the principal auth stored, or Anonymous when no auth
// layer ran.
func PrincipalFrom(ctx context.Context) Principal {
	if p, ok := ctx.Value(principalKey{}).(Principal); ok {
		return p
	}
	return Anonymous
}

// Target is the resource coordinate a call addressed, for the tools that take
// one. All four fields are emitted even when empty: AC1 wants the column to
// exist on every record, so an aggregation over kind sees "no coordinate"
// rather than a missing key it has to special-case.
type Target struct {
	APIVersion string
	Kind       string
	Namespace  string
	Name       string
}

// Result is how a call ended. Three values, no fourth (AC1): success, refused
// (the tool never ran — argument validation, the approval gate, an unknown
// name), error (the tool ran and reported failure).
type Result string

const (
	ResultSuccess Result = "success"
	ResultRefused Result = "refused"
	ResultError   Result = "error"
)

// Results is the closed set, for the same reason Reasons is one: the
// metrics of prd-metrics pre-register a series per value (AC1's "0 rather
// than absent"), and a list here cannot drift from the constants above.
var Results = []Result{ResultSuccess, ResultRefused, ResultError}

// Reason is why a refused call was refused (AC2, AC4).
type Reason string

const (
	ReasonAuthFailed       Reason = "auth_failed"
	ReasonInvalidInput     Reason = "invalid_input"
	ReasonUnconfigured     Reason = "unconfigured"
	ReasonGateRejected     Reason = "gate_rejected"
	ReasonGateExpired      Reason = "gate_expired"
	ReasonGateTimeout      Reason = "gate_timeout"
	ReasonGateUnreachable  Reason = "gate_unreachable"
	ReasonGateUnconfigured Reason = "gate_unconfigured"
)

// Reasons is the closed set, in prd-metrics AC2's order. It exists so the
// metrics of that PRD can pre-register every series from here rather than
// from a second list that can drift.
var Reasons = []Reason{
	ReasonAuthFailed, ReasonInvalidInput, ReasonUnconfigured,
	ReasonGateRejected, ReasonGateExpired, ReasonGateTimeout, ReasonGateUnreachable, ReasonGateUnconfigured,
}

// Gate is what the approval gate said about a gated call (AC2). Decision is
// the backend's verdict as the gate layer names it; it stays "approved" on a
// call the gate layer refused after the verdict (a target that moved, an
// approval already spent), which is exactly the case Reason and Decision are
// two fields for. AutoApproved is prd-approval-gate AC9's notice as a field.
type Gate struct {
	RequestID    string
	Decision     string
	AutoApproved bool
}

// Record is one tool call: AC1's five fields, plus AC2's reason and gate.
// The timestamp is not a field here because the sink stamps it at emission
// — a record built before the call and stamped after it would carry the
// wrong moment.
type Record struct {
	Tool      string
	Principal Principal
	Target    Target
	Result    Result
	Reason    Reason
	Gate      Gate

	// Duration is how long the dispatcher held the call with the gate wait
	// taken out, and GateWait how long the gate held it (prd-metrics AC3's
	// two histograms). Neither reaches the log line: the record's field set
	// is prd-event-log AC1/AC2's and closed, and "how long" is the metrics'
	// question. They ride on the record only because both surfaces derive
	// from this one emission point. Zero on a record the auth layer wrote —
	// that call was never timed because nothing of it ran.
	Duration time.Duration
	GateWait time.Duration
}

// Sink receives records. The dispatcher holds one so a test can capture
// records without scraping log output, and so the metrics of prd-metrics can
// later be derived at this same point rather than counted somewhere else
// (both PRDs' "표면 개요" require a single derivation point).
type Sink interface {
	Emit(ctx context.Context, r Record)
}

// Fanout is a Sink that hands each record to every sink it holds, in
// order. It is how the log line and the metrics of prd-metrics derive from
// the one emission point both PRDs require without the dispatcher knowing
// there are two of them.
type Fanout []Sink

// Emit implements Sink.
func (f Fanout) Emit(ctx context.Context, r Record) {
	for _, s := range f {
		s.Emit(ctx, r)
	}
}

// Log is the Sink that writes each record as one structured slog line on the
// default logger (AC5's stdout baseline). One line per record: the message is
// constant so a collector can select records by it, and everything that
// varies is an attribute.
type Log struct{}

// Message is the constant message of every record line.
const Message = "tool call"

// Emit implements Sink.
func (Log) Emit(ctx context.Context, r Record) {
	slog.Default().LogAttrs(ctx, slog.LevelInfo, Message, Attrs(r)...)
}

// Attrs renders a record as slog attributes, in the field order AC1 lists
// them (time is the handler's own attribute). The reason and gate keys are
// present on every record, empty on a success, for the same reason the
// target keys are.
func Attrs(r Record) []slog.Attr {
	return []slog.Attr{
		slog.String("tool", r.Tool),
		slog.String("principal", r.Principal.String()),
		slog.Group("target",
			slog.String("apiVersion", r.Target.APIVersion),
			slog.String("kind", r.Target.Kind),
			slog.String("namespace", r.Target.Namespace),
			slog.String("name", r.Target.Name),
		),
		slog.String("result", string(r.Result)),
		slog.String("reason", string(r.Reason)),
		slog.Group("gate",
			slog.String("request_id", r.Gate.RequestID),
			slog.String("decision", r.Gate.Decision),
			slog.Bool("auto_approved", r.Gate.AutoApproved),
		),
	}
}

File: tests/integration/_eventlog.py
"""이벤트 레코드(``msg="tool call"``)를 파드 stdout 에서 집는 공유 표면 (매칭 단위가 아니다).

`test-event-log.md` 의 시나리오 파일들이 같은 세 동작을 되풀이한다 — 배포의 로그 전량에서
레코드 줄만 고르고, 한 줄을 필드 사전으로 풀고, 호출 뒤 그 줄이 kubelet 의 로그 읽기에
나타날 때까지 기다린다. ``event_log_ac3_*.py`` 두 파일은 각자 들고 있었고 셋째 파일부터는
한 벌만 두는 편이 값이 어긋날 자리를 없앤다(``_gatekeeper.py`` 가 승인 조작을 한 벌만 두는
것과 같은 이유).

레코드 한 줄은 ``internal/eventlog/eventlog.go::Attrs`` 가 정한 순서의 ``key=value`` 열이고
그룹 필드는 ``target.kind``·``gate.decision`` 처럼 점으로 이어진다. 빈 값은 slog 의 text
핸들러가 ``key=""`` 로 낸다 — 그래서 성공 레코드의 ``reason`` 은 **없는 키가 아니라 빈 값**이고,
``parse`` 는 그것을 빈 문자열로 돌려준다(AC1 의 「필드 자체는 존재한다」를 그대로 보존).
"""

from __future__ import annotations

import re
import subprocess
import time

RECORD_MARK = 'msg="tool call"'

#: 응답이 돌아온 뒤 kubelet 의 로그 읽기가 그 줄을 보여 주기까지의 창.
RECORD_BUDGET = 15.0

_FIELD_RE = re.compile(r'(?P<key>[A-Za-z_][A-Za-z0-9_.]*)=(?P<value>"(?:[^"\\]|\\.)*"|\S*)')


def records(namespace: str, deployment: str) -> list[str]:
    """배포 파드 stdout 의 레코드 줄 전부(오래된 것부터)."""
    log = subprocess.check_output(
        ["kubectl", "-n", namespace, "logs", f"deploy/{deployment}", "--tail=-1"],
        text=True,
    )
    return [line for line in log.splitlines() if RECORD_MARK in line]


def parse(line: str) -> dict[str, str]:
    fields: dict[str, str] = {}
    for match in _FIELD_RE.finditer(line):
        value = match.group("value")
        if value.startswith('"'):
            value = value[1:-1].encode().decode("unicode_escape")
        fields[match.group("key")] = value
    return fields


def select(lines: list[str], **wanted: str) -> list[str]:
    """``tool="resource_patch", **{"target.name": "x"}`` 처럼 필드 값이 전부 일치하는 줄."""
    return [line for line in lines if all(parse(line).get(k) == v for k, v in wanted.items())]


def wait_for_new(namespace: str, deployment: str, before: int, **wanted: str) -> list[str]:
    """``wanted`` 에 맞는 레코드가 ``before`` 개보다 많아질 때까지 기다려 그 줄들을 돌려준다.

    예산 안에 늘지 않으면 그때의 줄들을 그대로 돌려주고 단언은 호출자가 한다 — 「호출당 정확히
    하나」는 이 헬퍼가 아니라 각 파일의 단언이다.
    """
    deadline = time.monotonic() + RECORD_BUDGET
    while True:
        found = select(records(namespace, deployment), **wanted)
        if len(found) > before or time.monotonic() >= deadline:
            return found
        time.sleep(0.5)


def exactly_one_new(found: list[str], before: int, what: str) -> dict[str, str]:
    assert len(found) - before == 1, (
        f"{what} 의 레코드가 {before} → {len(found)} — 호출당 정확히 하나여야 한다"
    )
    return parse(found[-1])
